apps/speed.c

```diff
@@ -561,52 +639,514 @@ static OPT_PAIR ecdh_choices[] = {
+        ret = DSA_verify(EVP_PKEY_DSA, buf, 20, buf2, kk, dsa_key[j]);
+        if (ret <= 0) {
+            BIO_printf(bio_err, "DSA verify failure\n");
+            ERR_print_errors(bio_err);
+            count = 1;
+            break;
+        }
+    }
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_EC
+static unsigned char ecdsasig[256];
+static unsigned int ecdsasiglen;
+static EC_KEY *ecdsa[EC_NUM];
+static long ecdsa_c[EC_NUM][2];
+static int ECDSA_sign_loop(void *args) {
+    int ret, count;
+    for (count = 0; COND(ecdsa_c[j][0]); count++) {
+        ret = ECDSA_sign(0, buf, 20,
+                ecdsasig, &ecdsasiglen, ecdsa[j]);
+        if (ret == 0) {
+            BIO_printf(bio_err, "ECDSA sign failure\n");
+            ERR_print_errors(bio_err);
+            count = 1;
+            break;
+        }
+    }
+    return count;
+}
+
+static int ECDSA_verify_loop(void *args) {
+    int ret, count;
+    for (count = 0; COND(ecdsa_c[j][1]); count++) {
+        ret = ECDSA_verify(0, buf, 20, ecdsasig, ecdsasiglen,
+                ecdsa[j]);
+        if (ret != 1) {
+            BIO_printf(bio_err, "ECDSA verify failure\n");
+            ERR_print_errors(bio_err);
+            count = 1;
+            break;
+        }
+    }
+    return count;
+}
+
+static unsigned char secret_a[MAX_ECDH_SIZE], secret_b[MAX_ECDH_SIZE];
+static EC_KEY *ecdh_a[EC_NUM], *ecdh_b[EC_NUM];
+static int outlen;
+static void *(*kdf) (const void *in, size_t inlen, void *out,
+        size_t *xoutlen);
+
+static int ECDH_compute_key_loop(void *args) {
+    int count;
+    for (count = 0; COND(ecdh_c[j][0]); count++) {
+        ECDH_compute_key(secret_a, outlen,
+                EC_KEY_get0_public_key(ecdh_b[j]),
+                ecdh_a[j], kdf);
+    }
+    return count;
+}
+#endif
+
+
+static int run_benchmark(int async_jobs, loopargs **array_loopargs, int (*loop_function)(void *)) {
+    int job_op_count = 0;
+    int total_op_count = 0;
+    int num_inprogress = 0;
+    int error = 0;
+    int i = 0;
+
+    run = 1;
+
+    if (0 == async_jobs) {
+        return loop_function((void *)array_loopargs[0]);
+    }
+
+    for (i = 0; i < async_jobs && !error; i++) {
+        switch (ASYNC_start_job(&(array_loopargs[i]->inprogress_job), &job_op_count,
+                                loop_function, (void *)array_loopargs[i], sizeof(loopargs))) {
+            case ASYNC_PAUSE:
+                ++num_inprogress;
+                break;
+            case ASYNC_FINISH:
+                if (job_op_count == -1) {
+                    error = 1;
+                } else {
+                    total_op_count += job_op_count;
+                }
+                break;
+            case ASYNC_NO_JOBS:
+            case ASYNC_ERR:
+                BIO_printf(bio_err, "Failure in the job\n");
+                ERR_print_errors(bio_err);
+                error = 1;
+                break;
+        }
+    }
+
+    while (num_inprogress > 0) {
```

Comment:
You are right, the windows version does busy waiting but this should not reduce the performance measured in the context of speed.

The idea is that if we are *burning* CPU cycles in the loop it means that the job is still in progress (e.g. the request to the HW accelerator is still in flight). In this case there would not have been much else to do for the CPU, hence it does not have a negative impact on the performance.